File: C++Solutions/SpiralMatrix_54.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> result;

        while(!matrix.empty()) {
            //append first row
            result.insert(result.end(),matrix.front().begin(), matrix.front().end());
            matrix.erase(matrix.begin());

            //add last element in each remaining array.
            for(int i = 0; i < matrix.size(); i++){
                if(!matrix[i].empty()){
                  result.push_back(matrix[i].back());
                  matrix[i].pop_back();
                }
            }

            //add the bottom array in reverse order
               if (!matrix.empty()) {
                for (int i = matrix.back().size() - 1; i >= 0; i--) {
                    result.push_back(matrix.back()[i]);
                }
                matrix.pop_back();
            }

            //add the first element of the remaining rows in reverse order
                for (int i = matrix.size() - 1; i >= 0; i--) {
                if (!matrix[i].empty()) {
                    result.push_back(matrix[i].front());
                    matrix[i].erase(matrix[i].begin());
                }
            }
        }

        return result;
    }
};

int main(){
    Solution solution;

    vector<vector<int>> matrix = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    vector<int> result = solution.spiralOrder(matrix);

    for(int num : result){
        cout << num << " ";
    }

    return 0;
}

File: JavaScriptSolutions/SpiralMatrix_54.js
const spiralMatrix = (matrix) => {
  let result = [];

  while (matrix.length > 0) {
    // Step 1: Add the first row into the result
    result = result.concat(matrix.shift());

    // Step 2: Pop off the last element in each remaining array
    if (matrix.length > 0) {
      for (let i = 0; i < matrix.length; i++) {
        if (matrix[i].length > 0) {
          // Check if there are elements left
          result.push(matrix[i].pop());
        }
      }
    }

    // Step 3: Pop the last row into the result in reverse order
    if (matrix.length > 0) {
      result = result.concat(matrix.pop().reverse());
    }

    // Step 4: Add the remaining first elements from each row
    if (matrix.length > 0) {
      for (let i = matrix.length - 1; i >= 0; i--) {
        if (matrix[i].length > 0) {
          // Check if there are elements left
          result.push(matrix[i].shift());
        }
      }
    }
  }

  return result;
};

const matrix = [
  [1, 2, 3],
  [4, 5, 6],
  [7, 8, 9],
];

console.log(spiralMatrix(matrix));
